File: Base/eval.py
"""The file contains helpers for evaluating instructions where all the operands
are known/constant.

"""

from typing import List, Dict, Optional
from Base import ir
from Base import opcode_tab as o

# TODO: naive implementation -> needs a lot more scrutiny
_EVALUATORS_ALU = {
    o.ADD: lambda x, y: x + y,
    o.SUB: lambda x, y: x - y,
    o.MUL: lambda x, y: x * y,
    o.DIV: lambda x, y: x // y,
    o.SHL: lambda x, y: x << y,
    o.OR: lambda x, y: x | y,
    o.AND: lambda x, y: x & y,
    o.XOR: lambda x, y: x ^ y,
}

_EVALUATORS_COND_BRA = {
    o.BEQ: lambda x, y: x == y,
    o.BNE: lambda x, y: x != y,
    o.BLT: lambda x, y: x < y,
    o.BLE: lambda x, y: x <= y,
}


def Cnttz(kind: o.DK, val: int):
    if val == 0:
        return kind.bitwidth()
    n = 0
    while val & 1 == 0:
        val >>= 1
        n += 1
    return n


def Cntlz(kind: o.DK, val: int):
    if val == 0:
        return kind.bitwidth()
    mask = 1 << (kind.bitwidth() - 1)
    n = 0
    while val & mask:
        mask >>= 1
        n += 1
    return n


_EVALUATORS_ALU1 = {
    o.CNTLZ: Cntlz,
    o.CNTTZ: Cnttz,
}


def EvaluatateALU(opcode: o.Opcode, op1: ir.Const, op2: ir.Const) -> ir.Const:
    evaluator = _EVALUATORS_ALU.get(opcode)
    assert evaluator, f"Evaluator NYI for: {opcode}"
    return ir.Const(op1.kind, evaluator(op1.value, op2.value))


def EvaluatateALU1(opcode: o.Opcode, op: ir.Const) -> Optional[ir.Const]:
    evaluator = _EVALUATORS_ALU1.get(opcode)
    assert evaluator, f"Evaluator NYI for: {opcode}"
    return ir.Const(op.kind, evaluator(op.kind, op.value))


def EvaluatateCondBra(opcode: o.Opcode, op1: ir.Const, op2: ir.Const) -> bool:
    evaluator = _EVALUATORS_COND_BRA.get(opcode)
    assert evaluator, f"Evaluator NYI for: {opcode}"
    return evaluator(op1.value, op2.value)


def ConvertIntValue(kind_dst: o.DK, val: ir.Const) -> ir.Const:
    kind_src = val.kind
    width_dst = kind_dst.bitwidth()
    width_src = kind_src.bitwidth()
    # print ("@@@", kind_dst.name, width_dst, kind_src, width_src, num_kind, x)
    masked = val.value & ((1 << width_dst) - 1)
    if width_dst > width_src:
        if kind_dst.flavor() == kind_src.flavor() or kind_src.flavor() == o.DK_FLAVOR_U:
            return ir.Const(kind_dst, val.value)
        # kind_dst == RK_U, kind_src == RK_S
        return ir.Const(kind_dst, masked)
    elif kind_dst.flavor() == o.DK_FLAVOR_U:
        return ir.Const(kind_dst, masked)
    else:
        # kind_dst[0] == RK_S
        sign = val.value & (1 << (width_dst - 1))
        if sign == 0:
            return ir.Const(kind_dst, masked)
        return ir.Const(kind_dst, masked - (1 << width_dst))

File: Base/eval.cc
// (c) Robert Muth - see LICENSE for more info

#include "Base/eval.h"
#include "Base/opcode_gen.h"
#include "Base/serialize.h"

namespace cwerg::base {
namespace {

template <typename VAL>
bool GenericEvaluateCondBra(OPC opc, VAL a, VAL b) {
  switch (opc) {
    case OPC::BEQ:
      return a == b;
    case OPC::BNE:
      return a != b;
    case OPC::BLE:
      return a <= b;
    case OPC::BLT:
      return a < b;
    default:
      ASSERT(false, "");
      return false;
  }
}

template <typename VAL>
VAL GenericEvaluateALU(OPC opc, DK dk, VAL va, VAL vb) {
  switch (opc) {
    case OPC::ADD:
      return va + vb;
    case OPC::SUB:
      return va - vb;
    case OPC::MUL:
      return va * vb;
    case OPC::DIV:
      return vb == 0 ? 0 : va / vb;
    case OPC::OR:
      return va | vb;
    case OPC::AND:
      return va & vb;
    case OPC::XOR:
      return va ^ vb;
      // TODO: REM
    case OPC::SHL:
      return va << (vb % DKBitWidth(dk));
    case OPC::SHR:
      return va >> (vb % DKBitWidth(dk));
    default:
      ASSERT(false, "unimplemented eval for " << EnumToString(opc));
      return 0;  // invalid
  }
}

template <>
double GenericEvaluateALU<double>(OPC opc, DK dk, double va, double vb) {
  switch (opc) {
    case OPC::ADD:
      return va + vb;
    case OPC::SUB:
      return va - vb;
    case OPC::MUL:
      return va * vb;
    case OPC::DIV:
      return vb == 0.0 ? 0 : va / vb;
    default:
      ASSERT(false, "unimplemented eval for " << EnumToString(opc));
      return 0;  // invalid
  }
}

}  // namespace

bool EvaluateCondBra(OPC opc, Const a, Const b) {
  switch (DKFlavor(ConstKind(a))) {
    case DK_FLAVOR_A:
    case DK_FLAVOR_C:
    case DK_FLAVOR_S: {
      return GenericEvaluateCondBra<int64_t>(opc, ConstValueACS(a),
                                             ConstValueACS(b));
    }
    case DK_FLAVOR_U: {
      return GenericEvaluateCondBra<uint64_t>(opc, ConstValueU(a),
                                              ConstValueU(b));
    }
      // TODO: DK_FLAVOR_F:
    default:
      ASSERT(false, "not supported");
      return false;
  }
}

Const EvaluateALU(OPC opc, Const a, Const b) {
  ASSERT(a.kind() == b.kind(), "");
  if (DKFlavor(ConstKind(a)) == DK_FLAVOR_U) {
    return ConstNewU(ConstKind(a),
                     GenericEvaluateALU<uint64_t>(
                         opc, ConstKind(a), ConstValueU(a), ConstValueU(b)));
  } else if (DKFlavor(ConstKind(a)) == DK_FLAVOR_F) {
    return ConstNewF(ConstKind(a),
                     GenericEvaluateALU<double>(
                         opc, ConstKind(a), ConstValueF(a), ConstValueF(b)));
  } else {
    return ConstNewACS(ConstKind(a), GenericEvaluateALU<int64_t>(
                                         opc, ConstKind(a), ConstValueACS(a),
                                         ConstValueACS(b)));
  }
}

Const EvaluateALU1(OPC opc, Const a) {
  ASSERT(false, "unimplemented eval for " << EnumToString(opc));
  return Const();
}

uint64_t ConstIntValue(Const src) {
  if (DKFlavor(ConstKind(src)) == DK_FLAVOR_U) {
    return ConstValueU(src);
  } else {
    return ConstValueACS(src);
  }
}

Const ConstWithUpdateKind(DK kind_dst, Const src) {
  if (DKFlavor(kind_dst) == DK_FLAVOR_U) {
    return ConstNewU(kind_dst, ConstIntValue(src));
  } else {
    return ConstNewACS(kind_dst, ConstIntValue(src));
  }
}

Const ConvertIntValue(DK kind_dst, Const src) {
  const DK kind_src = ConstKind(src);
  const uint64_t width_dst = DKBitWidth(kind_dst);
  const uint64_t width_src = DKBitWidth(kind_src);
  uint64_t mask = ~(-1LL << DKBitWidth(kind_dst));

  if (DKFlavor(kind_dst) == DK_FLAVOR_U) {
    return ConstNewU(kind_dst, mask & ConstIntValue(src));
  } else if (width_dst > width_src) {
    return ConstWithUpdateKind(kind_dst, src);
  } else {
    uint64_t v = ConstIntValue(src);
    // longer val to shorter signed
    int64_t will_be_negative = v & (1ULL << (width_dst - 1));
    if (will_be_negative) {
      return ConstNewACS(kind_dst, (v & mask) - (1ULL << width_dst));
    } else {
      return ConstNewACS(kind_dst, v & mask);
    }
  }
}

}  // namespace cwerg::base
